Update slots to reference before_render (#1493)
github/view_component: docs/guide/slots.md

<% end %>
```

## Referencing slots

> As the content passed to slots is registered after a component is initialized, it referenced in an initializer. One way to reference slots is using the `before_render` [lifecycle method](./lifecycle.md):

Reviewer: ```suggestion
As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](./lifecycle.md):
```

> As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](./lifecycle.md):

Reviewer: ```suggestion
As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](/guide/lifecycle):
```

> As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](/guide/lifecycle):

Reviewer 2: ```suggestion
As the content passed to slots is registered after a component is initialized, it cannot be referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](/guide/lifecycle):
```

```ruby
# blog_component.rb
class BlogComponent < ViewComponent::Base
  renders_one :image
  renders_many :posts

  def before_render

>     "PostContainer--hasImage" if image.present?

Reviewer: ```suggestion
    @post_container_classes = "PostContainer--hasImage" if image.present?
```

<% end %>
```

## Referencing slots

As the content passed to slots is registered after a component is initialized, it referenced in an initializer. One way to reference slots is using the `before_render` [lifecycle method](./lifecycle.md):

```ruby
# blog_component.rb
class BlogComponent < ViewComponent::Base
  renders_one :image
  renders_many :posts

  def before_render
    "PostContainer--hasImage" if image.present?
  end
end
```

```erb
<%# blog_component.html.erb %>
<% posts.each do |post| %>
  <div class="<%= @post_container_classes %>">

>     <% if image? %>
>       <%= image %>
>     <% end %>

Reviewer: ```suggestion
    <%= image if image? %>
```

>     <% if image? %>
>       <%= image %>
>     <% end %>

Reviewer: ```suggestion
    <%= image if image? %>
```

<% end %>
```

## Referencing slots

> You can access the instance of a slot in your component class by using the `before_render` [lifecycle method](./lifecycle.md).

Reviewer: ```suggestion
As slots are registered after initialization, use the `before_render` [lifecycle method](./lifecycle.md) to reference them:
```

```ruby

> # blog_component.rb
> class BlogComponent < ViewComponent::Base
>   renders_one :header

Reviewer: ```suggestion
# blog_component.rb
class BlogComponent < ViewComponent::Base
```

  renders_one :image
  renders_many :posts

>   def initialize(classes: [])
>     @post_container_classes = class_names("post-container", classes)
>     )
>   end
> 

Reviewer: ```suggestion
```

  def before_render

>     @post_container_classes = class_names(
>       {
>         "PostContainer--hasImage": image.present?
>       },
>       @post_container_classes
>     )

Reviewer: ```suggestion
    "PostContainer--hasImage" if image.present?
```

  end
end
```

```erb
<%# blog_component.html.erb %>

> <h1><%= header %></h1>
> 

Reviewer: ```suggestion
```

<% posts.each do |post| %>

>   <div class="<%= @post_container_classes =%>">

Reviewer: ```suggestion
  <div class="<%= @post_container_classes %>">
```

<% end %>
```

## Referencing slots

> You can access the instance of a slot in your ruby file by using the `before_render` [lifecycle method](./lifecycle.md).

Reviewer 2: ```suggestion
You can access the instance of a slot in your component class by using the `before_render` [lifecycle method](./lifecycle.md).
```

Author: Good suggestion, I always trip over this type of terminology :)

```ruby
# blog_component.rb
class BlogComponent < ViewComponent::Base
  renders_one :header
  renders_one :image
  renders_many :posts

>   def initialize(**system_arguments)
>     @system_arguments = system_arguments
> 
>     @post_container_classes = class_names(
>       "PostContainer",
>       system_arguments[:classes]

Reviewer 2: Can we simplify a bit by removing `system_arguments`?

```suggestion
  def initialize(classes: [])
    @post_container_classes = class_names("post-container", classes)
```

What do you think?

Author: Absolutely!
